### PR TITLE
Partial documentation

### PR DESCRIPTION
**Pull request by tomas-campodonico**

### CODE REVIEW

**Comment by krosti** [comment]
+1

File: README.md [hunk 1]
```diff
@@ -32,3 +28,152 @@
   1. Socket.io and Redis
   2. Popular modules
   3. Links
+
+## Introduction to NodeJS ##
+### CommonJS Modules ###
+Code in node apps is organized in modules. There are a couple of standards on how to write a module in JavaScript. By convention, NodeJS has decided to use the CommonJS Module Specification.
```
Comment (eabait) [inline README.md]: NodeJS application organize their code into modules. There are a couple of standards on how to write a module in JavaScript. By convention, NodeJS has decided to use the CommonJS Module Specification.

File: README.md [hunk 2]
```diff
@@ -32,3 +28,152 @@
   1. Socket.io and Redis
   2. Popular modules
   3. Links
+
+## Introduction to NodeJS ##
```
Comment (eabait) [inline README.md:34]: In this section, you will learn the basics of NodeJS. From how NodeJS application organize their source code, to how application dependencies are handled by NPM, and which global objects are provided by default.

File: README.md [hunk 3]
```diff
@@ -32,3 +28,152 @@
...
+### CommonJS Modules ###
+Code in node apps is organized in modules. There are a couple of standards on how to write a module in JavaScript. By convention, NodeJS has decided to use the CommonJS Module Specification.
+
+* [Module 1.0 - CommonJS Specification](http://www.commonjs.org/specs/modules/1.0/ "Module 1.0 - CommonJS Specification")
+* [A complete explanation of the 3 most used module definitions](http://addyosmani.com/writing-modular-js/ "Modules in JavaScript")
+### Intro to npm ###
```
Comment (eabait) [inline README.md]: Change npm to NPM.

File: README.md [hunk 4]
```diff
@@ -32,3 +28,152 @@
...
+npm is a package manager for JavaScript, and is the default for Node.js. npm makes it easy for JavaScript developers to share and reuse code, and it makes it easy to update the code that you're sharing.
+
+* [What is npm? Short explanation video](https://www.youtube.com/watch?v=pa4dc480Apo)
+* [Installing node and updating npm](https://docs.npmjs.com/getting-started/installing-node)
+* [Create and publish modules to npm](https://quickleft.com/blog/creating-and-publishing-a-node-js-module/)
+
+### Globals, Process and Buffer###
+As we have learnt, node's module system discourages the use of globals; however node provides a few important globals for use to utilize. Here we list the most important ones.
```
Comment (eabait) [inline README.md]: As we have learnt, NodeJS's module system discourages the use of globals. However, it provides a few important globals that we can use. This section list the most important ones.

File: README.md [hunk 5]
```diff
@@ -32,3 +28,152 @@
...
+
+- [Buffer](http://nodejs.org/api/buffer.html): To handle binary data, node provides us with the global Buffer object. Buffer instances represent memory allocated independently of V8's heap. [Here](http://www.w3resource.com/node.js/nodejs-buffer.php) is an example of how to use buffers.
```
Comment (eabait) [inline README.md]: Change "node provides us with the global Buffer object" to "NodeJS provides the global Buffer object"

File: README.md [hunk 6]
```diff
@@ -32,3 +28,152 @@
...
+
+### Globals, Process and Buffer###
+As we have learnt, node's module system discourages the use of globals; however node provides a few important globals for use to utilize. Here we list the most important ones.
+
+- [Buffer](http://nodejs.org/api/buffer.html): To handle binary data, node provides us with the global Buffer object. Buffer instances represent memory allocated independently of V8's heap. [Here](http://www.w3resource.com/node.js/nodejs-buffer.php) is an example of how to use buffers.	 
+- [Global](http://nodejs.org/api/globals.html): This module represent the global scope. It is not a good practice to use this global. When you define a global variable in a browser, its scope is the global one, but in node is different: you define a variable inside a module without "var" and its scope is the module itself.
+- [Process](http://nodejs.org/api/process.html): Process is a node module that can be accessed anywhere without the need to be required. It exposes process manipulation such as signalling, exiting, the process id (pid), and more.
+## Asynchronous Patterns ##
```
Comment (eabait) [inline README.md:55]: This section introduces NodeJS's event loop, and different patterns to deal with asynchronous control-flow.

File: README.md [hunk 7]
```diff
@@ -32,3 +28,152 @@
...
+- [Using async functions in practice](http://www.sebastianseilund.com/nodejs-async-in-practice)
+- [Async iteration patterns](http://metaduck.com/01-asynchronous-iteration-patterns.html)
+
+### Control Flow and Exception Handling ###
+- [Solutions to control the flow](http://book.mixu.net/node/ch7.html)
+- [Managing the asynchronous nature of NodeJS](http://code.tutsplus.com/tutorials/managing-the-asynchronous-nature-of-nodejs--net-36183)
+- [The Async library](https://github.com/caolan/async)
+- [Error Handling in NodeJS](https://www.joyent.com/developers/node/design/errors)
+
+## Communication ##
```
Comment (eabait) [inline README.md:73]: This section introduces the use of events, clusters, and streams in NodeJS.

File: README.md [hunk 8]
```diff
@@ -32,3 +28,152 @@
...
+Many objects in Node emit events: a net.Server emits an event each time a peer connects to it, a fs.readStream emits an event when the file is opened. All objects which emit events are instances of events.EventEmitter.
+[http://nodejs.org/api/events.html](http://nodejs.org/api/events.html)
+
+### Cluster ###
+A single instance of Node runs in a single thread. To take advantage of multi-core systems the user will sometimes want to launch a cluster of Node processes to handle the load. [http://nodejs.org/api/cluster.html](http://nodejs.org/api/cluster.html)
+
+### Streams ###
+A stream is an abstract interface implemented by various objects in Node. Streams are pipes that let you easily read data from a source and pipe it to a destination. [http://nodejs.org/api/stream.html](http://nodejs.org/api/stream.html): 
+
+## RESTful API ##
```
Comment (eabait) [inline README.md:96]: This section introduces RESTful principles, along with the HTTP protocol basics.

File: README.md [hunk 9]
```diff
@@ -32,3 +28,152 @@
...
+## MongoDB ##
+MongoDB is an open source document-oriented database that provides high availability, high performance and automatic scaling.
+
+- [Official documentation](http://docs.mongodb.org/manual/)
+- [MongoDB architecture: Data Model, Query Model and Management](http://www.mongodb.com/mongodb-architecture)
+- [Performance best practices for MongoDB](http://info.mongodb.com/rs/mongodb/images/MongoDB-Performance-Best-Practices.pdf)
+- [Tutorial: creating an API with Node and MongoDB](http://adrianmejia.com/blog/2014/10/01/creating-a-restful-api-tutorial-with-nodejs-and-mongodb/)
+
+
+## Debugging and Profiling ##
```
Comment (eabait) [inline README.md:143]: In this section, you will learn how to debug, and profile a NodeJS application.

File: README.md [hunk 10]
```diff
@@ -32,3 +28,152 @@
...
+- [How to debug using Node inspector](https://vimeo.com/77870960)
+- [Debugging apps using WebStorm](https://www.jetbrains.com/webstorm/help/running-and-debugging-node-js.html)
+- [Profiling NodeJS applications](http://www.willvillanueva.com/the-node-js-profiling-guide-that-hasnt-existed-profiling-node-js-applications-part-1/)
+- [CPU Profiling](http://www.clock.co.uk/blog/easy-cpu-profiling-in-node-js)
+
+## Others ##
+### Socket.io ###
```
Comment (eabait) [inline README.md]: Question: Shall we move Socket.io to Communication section?

File: README.md [hunk 11]
```diff
@@ -32,3 +28,152 @@
...
+ Socket.IO enables real-time bidirectional event-based communication.
+It works on every platform, browser or device, focusing equally on reliability and speed.
+
+
+* [Engine.IO: the realtime engine](https://github.com/Automattic/engine.io)
+* [Socket.IO](http://socket.io/docs/)
+* [Migration from 0.9 to 1.0](http://socket.io/docs/migrating-from-0-9/)
+* [Example application](https://github.com/Automattic/socket.io/tree/master/examples/chat)
+
+### Redis ###
```
Comment (eabait) [inline README.md]: Lets group MongoDB, and Redis into a single section. We can call the section "Data Repositories" or something along that line

File: README.md [hunk 12]
```diff
@@ -32,3 +28,152 @@
...
+- [Tutorial: creating an API with Node and MongoDB](http://adrianmejia.com/blog/2014/10/01/creating-a-restful-api-tutorial-with-nodejs-and-mongodb/)
+
+
+## Debugging and Profiling ##
+- [How to debug using Node inspector](https://vimeo.com/77870960)
+- [Debugging apps using WebStorm](https://www.jetbrains.com/webstorm/help/running-and-debugging-node-js.html)
+- [Profiling NodeJS applications](http://www.willvillanueva.com/the-node-js-profiling-guide-that-hasnt-existed-profiling-node-js-applications-part-1/)
+- [CPU Profiling](http://www.clock.co.uk/blog/easy-cpu-profiling-in-node-js)
+
+## Others ##
```
Comment (eabait) [inline README.md:151]: This section presents relevant modules written by the community, which are proven, and useful to write any NodeJS application. In addition, we provide further documentation to for you to explore.

Comment (facundocabrera) [inline README.md:151]: http://www.nearform.com/nodecrunch/self-detect-memory-leak-node/

File: README.md (hunk 7 above)
Comment (Angelfire) [inline README.md:73]: https://github.com/rvagg/learnyounode
Also http://nodeschool.io/ is a better way to learn Nodejs, their CLI it's a great tool.

**Pull request merged**